### PR TITLE
ddl: fix reorg goroutine leak after column-type-change is cancelled

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/24584

### What is changed and how it works?

What's Changed:
When cancel action takes place in the reorg state of CTC, we should give the DDL one more round to fetch the `errCancelledDDLJob` from the bottom goroutine up. 

runReorgJob is used as a portal to do the reorganization work.
eg:
- 1: add index
- 2: alter column type
- 3: clean global index
```
// ddl goroutine >---------+
//   ^                     |
//   |                     |
//   |                     |
//   |                     | <---(doneCh)--- f()
// HandleDDLQueue(...)     | <---(regular timeout)
//   |                     | <---(ctx done)
//   |                     |
//   |                     |
// A more ddl round  <-----+
```
How can we cancel reorg job?

The background reorg is continuously running except for several factors, for instances, ddl owner change,
logic error (kv duplicate when insert index / cast error when alter column), ctx done, and cancel signal.

When `admin cancel ddl jobs xxx` takes effects, we will give the this kind of reorg ddl one more round.
because we should pull the result from doneCh out, otherwise, the reorg worker will hang on `f()` logic,
which is a kind of goroutine leak.

That's why we couldn't set the job to rollingback state directly in `convertJob2RollbackJob`, which is a
cancelling portal for admin cancel action.

In other words, the cancelling signal is informed from the bottom up, we set the atomic cancel variable
in the cancelling portal to notify the lower worker goroutine, and fetch the cancel error from them in
the additional ddl round.

After that, we can make sure that the worker goroutine is correctly shut down.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- ddl: fix reorg goroutine leak after column-type-change is cancelled

### CODE REVIEW

File: ddl/column_type_change_test.go [hunk 1]
```diff
@@ -1797,3 +1799,59 @@ func (s *testColumnTypeChangeSuite) TestChangeIntToBitWillPanicInBackfillIndexes
 		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin"))
 	tk.MustQuery("select * from t").Check(testkit.Rows("\x13 1 1.00", "\x11 2 2.00"))
 }
+
+// Close issue #24584
+func (s *testColumnTypeChangeSuite) TestCancelCTCInReorgStateWillCauseGoroutineLeak(c *C) {
```
Comment (tangenta) [inline ddl/column_type_change_test.go:1804]: Has this test ever failed before you change `ddl/reorg.go`?

Comment (AilinKid) [reply ddl/column_type_change_test.go:1804]: because the previous logic won't notify the background worker(notifyReorgCancel)

For the old code, cancel can always succeed, because it doesn't have to notify the worker. (it is still running)
For the test here, we mock an infinite worker here, the cancel succeeds only if the reorg exit.

File: ddl/reorg.go [hunk 2]
```diff
@@ -155,6 +155,40 @@ func (rc *reorgCtx) clean() {
...
+//   |                     |
+//   |                     |
+// A more ddl round  <-----+
+//
+// How can we cancel reorg job?
+//
+// The background reorg is continuously running except for several factors, for instances, ddl owner change,
+// logic error (kv duplicate when insert index / cast error when alter column), ctx done, and cancel signal.
+//
+// When `admin cancel ddl jobs xxx` takes effects, we will give the this kind of reorg ddl one more round.
```
Comment (djshow832) [inline ddl/reorg.go:180]:
```suggestion
// When `admin cancel ddl jobs xxx` takes effect, we will give this kind of reorg ddl one more round.
```

File: ddl/rollingback.go [hunk 3]
```diff
@@ -110,7 +110,15 @@ func convertNotStartAddIdxJob2RollbackJob(t *meta.Meta, job *model.Job, occuredE
 // Since modifying column job has two types: normal-type and reorg-type, we should handle it respectively.
 // normal-type has only two states:    None -> Public
 // reorg-type has five states:         None -> Delete-only -> Write-only -> Write-org -> Public
-func rollingbackModifyColumn(t *meta.Meta, job *model.Job) (ver int64, err error) {
+func rollingbackModifyColumn(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	// If the value of SnapshotVer isn't zero, it means the reorg workers has been started.
```
Comment (djshow832) [inline ddl/rollingback.go:114]:
```suggestion
	// If the value of SnapshotVer isn't zero, it means the reorg workers have been started.
```

File: ddl/rollingback.go [hunk 4]
```diff
@@ -138,7 +146,7 @@ func rollingbackModifyColumn(t *meta.Meta, job *model.Job) (ver int64, err error
 		job.State = model.JobStateCancelled
 		return ver, errCancelledDDLJob
 	}
-	// The job has been in it's middle state and we roll it back.
+	// The job has been in it's middle state (but reorg worker hasn't start) and we rolling it back here.
```
Comment (djshow832) [inline ddl/rollingback.go:149]:
```suggestion
	// The job has been in its middle state (but the reorg worker hasn't started) and we roll it back here.
```

**Review by tangenta — APPROVED** [review]

**Review by djshow832 — APPROVED** [review]
